### PR TITLE
Fix more race conditions around `git4idea.branch.GitBrancher`

### PR DESCRIPTION
**Pull request by PawelLipski**
image

### CODE REVIEW

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/traverse/TraverseSyncToParent.java [hunk 1]
```diff
@@ -125,8 +125,8 @@ private void handleMergedToParent(
         // For a branch merged to its parent, we're not syncing to remote.
         // Let's just go straight to the next branch.
         Runnable doInUIThreadWhenReady = () -> graphTable.queueRepositoryUpdateAndModelRefresh(traverseNextEntry);
-        new SlideOutBackgroundable(getString("action.GitMachete.BaseSlideOutAction.task.title"),
-            managedBranch, gitRepository, currentBranchIfManaged, branchLayout, graphTable, doInUIThreadWhenReady);
```
Comment (PawelLipski) [inline frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/traverse/TraverseSyncToParent.java:129]: Whooops `queue()` wasn't even called here, slide out in traverse doesn't work 😅

Comment (PawelLipski) [reply frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/traverse/TraverseSyncToParent.java]: Resolved on #1471

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/base/BaseSyncToParentByMergeAction.java [hunk 2]
```diff
@@ -104,17 +101,8 @@ public void doMergeIntoCurrentBranch(GitRepository gitRepository, MergeProps mer
     log().debug(() -> "Entering: project = ${project}, gitRepository = ${gitRepository}," +
         " stayingBranch = ${stayingBranch}");
 
-    new Task.Backgroundable(project, getString("action.GitMachete.BaseSyncToParentByMergeAction.task-title.current")) {
```
Comment (PawelLipski) [inline frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/base/BaseSyncToParentByMergeAction.java:107]: Not a race condition, just a redundant backgroundable wrapper

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/SlideOut.java [hunk 3]
```diff
@@ -31,125 +32,100 @@
...
   private final String currentBranchNameIfManaged;
   private final BranchLayout branchLayout;
   private final GitRepository gitRepository;
-  private final BaseEnhancedGraphTable graphTable;
-  private final @UI Runnable doInUIThreadWhenReady;
 
   public static final String DELETE_LOCAL_BRANCH_ON_SLIDE_OUT_GIT_CONFIG_KEY = "machete.slideOut.deleteLocalBranch";
 
-  public SlideOutBackgroundable(String title, IManagedBranchSnapshot branchToSlideOut,
+  public SlideOut(IManagedBranchSnapshot branchToSlideOut,
```
Comment (PawelLipski) [inline frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/SlideOut.java:52]: The problem with the naming now... is that this class is still _kinda a backgroundable_, as part of the workload (deleting the branch, getting/setting config) happens in backgroundables 😅 but this should be apparent from `@ContinuesInBackground` on `run`

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/SlideOut.java [hunk 4]
```diff
@@ -31,125 +32,100 @@
...
       val root = gitRepository.getRoot();
-      val shouldDelete = getDeleteLocalBranchOnSlideOutGitConfigValue(root);
-      if (shouldDelete == null) {
-        ModalityUiUtil.invokeLaterIfNeeded(ModalityState.NON_MODAL,
-            () -> suggestBranchDeletion(branchToSlideOutName, doInUIThreadWhenReady));
-      } else {
-        handleBranchDeletionDecision(branchToSlideOutName, shouldDelete);
-        ModalityUiUtil.invokeLaterIfNeeded(ModalityState.NON_MODAL, doInUIThreadWhenReady);
-      }
+      getDeleteLocalBranchOnSlideOutGitConfigValueAndExecute(root, shouldDelete -> {
```
Comment (PawelLipski) [inline frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/SlideOut.java:79]: That's basically a [_callback hell_](http://callbackhell.com/).

Until IntelliJ (and Java in general) migrates to Futures, it won't get any easier ☹️ `map`/`flatMap` are waaaay easier to manipulate than this 💩 I didn't face even the tenth of race conditions, and didn't need to invent stuff like what's in `BackgroundTaskEnqueuingTestSuite`.

To absolve IntelliJ a bit, the fact is that the strict separation between UI and non-UI thread **might** make using Futures more cumbersome... I haven't ever seen such combination in practice. Mostly used Futures for server-side code, where still some amount of [bulkheading](https://stackoverflow.com/questions/47250947/what-is-bulkheading-in-computer-science) is present... but that doesn't come even close to the strictness of UI threading.

File: src/test/java/com/virtuslab/archunit/BackgroundTaskEnqueuingTestSuite.java [hunk 5]
```diff
@@ -146,7 +146,7 @@ public void check(JavaMethod method, ConditionEvents events) {
               String callTargetName = callTarget.getName();
               JavaClass callTargetOwner = callTarget.getOwner();
               if (callTargetOwner.isAssignableTo(Task.Backgroundable.class) && callTargetName.equals("queue") ||
-                  callTargetOwner.isAssignableTo(GitBrancher.class) && callTargetName.equals("checkout") ||
+                  callTargetOwner.isAssignableTo(GitBrancher.class) && !callTargetName.equals("getInstance") ||
```
Comment (PawelLipski) [inline src/test/java/com/virtuslab/archunit/BackgroundTaskEnqueuingTestSuite.java:149]: Here's the core change that spurred all other fixes in this PR (although in fact, what I've first seen was the race condition on deleting branch in slide out, which caused a _branch not found_ notification to display)

File: src/test/java/com/virtuslab/archunit/UIThreadUnsafeMethodInvocationsTestSuite.java [hunk 6]
```diff
@@ -139,6 +141,7 @@ public void check(JavaMethod method, ConditionEvents events) {
       "git4idea.branch.GitNewBranchDialog.showAndGetOptions()",
       "git4idea.branch.GitNewBranchOptions.getName()",
       "git4idea.branch.GitNewBranchOptions.shouldCheckout()",
+      "git4idea.config.GitConfigUtil.getBooleanValue(java.lang.String)",
```
Comment (PawelLipski) [inline src/test/java/com/virtuslab/archunit/UIThreadUnsafeMethodInvocationsTestSuite.java:144]: `getBooleanValue` is just a helper to convert string (typically retrieved via `GitConfigUtil.getValue`) to a `Boolean`, it doesn't really access git config

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/SlideOut.java [hunk 7]
```diff
@@ -173,35 +149,49 @@ private void slideOutBranch(String branchName) {
...
-      if (value != null) {
-        Boolean booleanValue = GitConfigUtil.getBooleanValue(value);
-        return booleanValue != null && booleanValue;
-      }
-    } catch (VcsException e) {
-      LOG.info(
-          "Attempt to get '${DELETE_LOCAL_BRANCH_ON_SLIDE_OUT_GIT_CONFIG_KEY}' git config value failed: key may not exist");
-    }
+  @ContinuesInBackground
+  private void getDeleteLocalBranchOnSlideOutGitConfigValueAndExecute(VirtualFile root,
```
Comment (PawelLipski) [inline frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/SlideOut.java:159]: I've even thought of forcing that onto UI thread (with some `@IgnoreUIThreadUnsafeCalls` to circumvent our own QA)... but got [brutally stopped by IntelliJ itself](https://app.circleci.com/pipelines/github/VirtusLab/git-machete-intellij-plugin/8199/workflows/e8365e4c-2bb1-4237-8b7a-26f86cfd8f99/jobs/8625/steps): 

```
  [intellij-stdout] java.lang.Throwable: Assertion failed: Should not wait for built-in server on EDT
    [intellij-stdout] 	at com.intellij.openapi.diagnostic.Logger.assertTrue(Logger.java:254)
    [intellij-stdout] 	at org.jetbrains.ide.BuiltInServerManagerImpl.waitForStart(BuiltInServerManagerImpl.kt:97)
    [intellij-stdout] 	at git4idea.rebase.GitRebaseEditorService.getEditorCommand(GitRebaseEditorService.java:75)
    [intellij-stdout] 	at git4idea.rebase.GitHandlerRebaseEditorManager.prepareEditor(GitHandlerRebaseEditorManager.java:42)
    [intellij-stdout] 	at git4idea.rebase.GitHandlerRebaseEditorManager.lambda$prepareEditor$0(GitHandlerRebaseEditorManager.java:28)
    [intellij-stdout] 	at git4idea.GitUtil.tryRunOrClose(GitUtil.java:1034)
    [intellij-stdout] 	at git4idea.rebase.GitHandlerRebaseEditorManager.prepareEditor(GitHandlerRebaseEditorManager.java:27)
    [intellij-stdout] 	at git4idea.commands.GitImplBase.prepareGeneralPurposeEditor(GitImplBase.java:178)
    [intellij-stdout] 	at git4idea.commands.GitImplBase.run(GitImplBase.java:162)
    [intellij-stdout] 	at git4idea.commands.GitImplBase.runCommand(GitImplBase.java:63)
    [intellij-stdout] 	at git4idea.config.GitConfigUtil.getValue(GitConfigUtil.java:80)
    [intellij-stdout] 	at git4idea.config.GitConfigUtil.getValue(GitConfigUtil.java:73)
    [intellij-stdout] 	at com.virtuslab.gitmachete.frontend.actions.common.SlideOut.getDeleteLocalBranchOnSlideOutGitConfigValue(SlideOut.java:158)
    [intellij-stdout] 	at com.virtuslab.gitmachete.frontend.actions.common.SlideOut.run(SlideOut.java:76)
    [intellij-stdout] 	at com.virtuslab.gitmachete.frontend.actions.base.BaseSlideOutAction.doSlideOut(BaseSlideOutAction.java:78)
    [intellij-stdout] 	at com.virtuslab.gitmachete.frontend.actions.base.BaseSlideOutAction.actionPerformed(BaseSlideOutAction.java:58)
    [intellij-stdout] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    [intellij-stdout] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
    [intellij-stdout] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    [intellij-stdout] 	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
....
```

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/SlideOut.java [hunk 8]
```diff
@@ -173,35 +149,49 @@ private void slideOutBranch(String branchName) {
...
+      Consumer<@Nullable Boolean> doForConfigValue) {
+    new Task.Backgroundable(project, getString("action.GitMachete.get-git-config.task-title")) {
 
-    return null;
+      @Override
+      @UIThreadUnsafe
+      public void run(ProgressIndicator indicator) {
+        Boolean result = null;
+        try {
+          val value = GitConfigUtil.getValue(project, root, DELETE_LOCAL_BRANCH_ON_SLIDE_OUT_GIT_CONFIG_KEY);
```
Comment (PawelLipski) [inline frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/SlideOut.java:168]: So yeah, this `getValue` must happen outside of EDT, period ☹️

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/SlideOut.java [hunk 9]
```diff
@@ -31,125 +32,110 @@
...
+import com.virtuslab.gitmachete.frontend.resourcebundles.GitMacheteBundle;
 import com.virtuslab.gitmachete.frontend.ui.api.table.BaseEnhancedGraphTable;
 import com.virtuslab.qual.async.ContinuesInBackground;
 import com.virtuslab.qual.guieffect.UIThreadUnsafe;
 
 @CustomLog
-// For some reason, using `@ExtensionMethod({GitMacheteBundle.class})` on this class
-// leads to weird Checker Framework errors :/
-public class SlideOutBackgroundable extends Task.Backgroundable {
+@ExtensionMethod({GitMacheteBundle.class})
```
Comment (mkondratek) [inline frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/SlideOut.java:41]: issues are not an issue no more 😅

**Review by mkondratek — APPROVED** [review]
nice. less backgroundables, more stability and better performance

**Pull request merged**